cmd/registry/patch/api.go

```diff
@@ -0,0 +1,196 @@
+	api.Data.DisplayName = message.DisplayName
+	api.Data.Description = message.Description
+	api.Data.Availability = message.Availability
+	api.Data.RecommendedVersion, err = relativeVersionName(apiName, message.RecommendedVersion)
+	if err != nil {
+		return nil, err
+	}
+	api.Data.RecommendedDeployment, err = relativeDeploymentName(apiName, message.RecommendedDeployment)
+	if err != nil {
+		return nil, err
+	}
+	err = core.ListVersions(ctx, client, apiName.Version(""), "", func(message *rpc.ApiVersion) {
+		version, err2 := newAPIVersion(ctx, client, message)
+		// unset these because they can be inferred
+		version.APIVersion = ""
+		version.Kind = ""
+		if err2 == nil {
+			api.Data.APIVersions = append(api.Data.APIVersions, version)
+		} else {
+			err = err2
+		}
+	})
+	if err != nil {
+		return nil, err
+	}
+	err = core.ListDeployments(ctx, client, apiName.Deployment(""), "", func(message *rpc.ApiDeployment) {
+		deployment, err2 := newAPIDeployment(ctx, client, message)
+		// unset these because they can be inferred
+		deployment.APIVersion = ""
+		deployment.Kind = ""
+		if err2 == nil {
+			api.Data.APIDeployments = append(api.Data.APIDeployments, deployment)
+		} else {
+			err = err2
+		}
+	})
+	return api, err
+}
+
+// ExportAPI allows an API to be individually exported as a YAML file.
+func ExportAPI(ctx context.Context, client *gapic.RegistryClient, message *rpc.Api) ([]byte, *Header, error) {
+	api, err := newAPI(ctx, client, message)
+	if err != nil {
+		return nil, nil, err
+	}
+	b, err := yaml.Marshal(api)
+	if err != nil {
+		return nil, nil, err
+	}
+	return b, &api.Header, nil
+}
+
+// relativeVersionName returns the version id if the version is within the specified API
+func relativeVersionName(apiName names.Api, version string) (string, error) {
+	if version == "" {
+		return "", nil
+	}
+	versionName, err := names.ParseVersion(version)
+	if err != nil {
+		return "", err
+	}
+	if versionName.Api().String() == apiName.String() {
+		return versionName.VersionID, nil
+	}
+	return version, nil
+}
+
+// relativeDeploymentName returns the deployment id if the deployment is within the specified API
+func relativeDeploymentName(apiName names.Api, deployment string) (string, error) {
+	if deployment == "" {
+		return "", nil
+	}
+	deploymentName, err := names.ParseDeployment(deployment)
+	if err != nil {
+		return "", err
+	}
+	if deploymentName.Api().String() == apiName.String() {
+		return deploymentName.DeploymentID, nil
+	}
+	return deployment, nil
+}
+
+// optionalVersionName returns a version name if the id is not empty
+func optionalVersionName(apiName names.Api, versionID string) string {
+	if versionID == "" {
+		return ""
+	}
+	return apiName.Version(versionID).String()
+}
+
+// optionalDeploymentName returns a deployment name if the id is not empty
+func optionalDeploymentName(apiName names.Api, deploymentID string) string {
+	if deploymentID == "" {
+		return ""
+	}
+	return apiName.Deployment(deploymentID).String()
+}
```

Comment:
These functions assume `deploymentID` is truly an ID, but there's there's no validation here or in the caller to prevent a full resource name from being passed in. Users who set their recommended version to a full resource name will end up with a malformed name being stored.

Reply:
I've addressed the need for validation with a TODO in the code.